### DISCUSSION TITLE
How to abort model creation based on response from external API?

### DISCUSSION THREAD

**Discussion by adevade** [filamentphp/filament#3842] [Help] [answered]
Kind of new to Filament, and tinkering around with a personal project to learn more.

I want to create a `Site` through Filament, where I give it a domain, `example.com`.
When creating, I want to check if this domain already exists in CloudFlare.

- If it does exists, I want to save data from the response to the model.
- If it does *not* exist, I want to first create it in CloudFlare and then save the data to model.

^^ This is the happy path. I got this under control.

But say that something goes wrong when contacting the API. An exception is thrown or some data is invalid.
I then want to abort creation of the model and show a notification that something went wrong. How would I handle that? 

Any and all insightful tips are greatly appreciated! ❤

-----

**EDIT: Clarification**

I'm also uncertain on where this logic should go. Before/after data validation? Overwriting the `handleRecordCreation()` method on the Create page component?

**Comment by danharrin** [comment, marked as answer]
Yeah, `handleRecordCreation()` is probably correct. You can throw a ValidationException from there if something goes wrong, that will halt execution. A validation error is probably better UX than a notification as well.

**Reply by adevade** [reply]
Thanks! It works like a charm!

If the request fails, I'm throwing a ValidationException with the following code:
```php
throw ValidationException::withMessages([
    'data.domain' => $response->json('errors.0.message', 'Error creating domain on Cloudflare.'),
]);
```

**Reply by danharrin** [reply]
Hey! There's a better way to do this now - #4433

**Comment by njxqlus** [comment]
For me the next example works. I used `mountedActionsData.0.field_name`
```php
TextInput::make('yt_id')
    ->required()
    ->label('YouTrack ID')
    ->live()
    ->suffixAction(
        Action::make('Get Data')
            ->action(function (Set $set, $state, ListDefects $livewire) {
                try {
                    $issue = IssueController::get($state);
                    $set('name', $issue->summary);
                    $set('state', $issue->state);
                } catch (Exception $e) {
                    throw ValidationException::withMessages([
                        'mountedActionsData.0.yt_id' => $e->getMessage(),
                    ]);
                }
            })
            ->icon('heroicon-o-arrow-down-tray')
            ->disabled(fn($state) => empty($state))
    ),
```
